src/Storages/ExternalStream/Kafka/Kafka.cpp

```diff
@@ -276,45 +363,185 @@ void Kafka::validateMessageKey(const String & message_key_, IStorage * storage,
+
+    // const auto & settings_ref = context->getSettingsRef();
+    /*auto share_resource_group = (settings_ref.query_resource_group.value == "shared") && (settings_ref.seek_to.value == "latest");
+    if (share_resource_group)
+    {
+        for (Int32 i = 0; i < shards; ++i)
+        {
+            if (!column_names.empty())
+                pipes.emplace_back(source_multiplexers->createChannel(i, column_names, metadata_snapshot, context));
+            else
+                pipes.emplace_back(source_multiplexers->createChannel(i, {RESERVED_APPEND_TIME}, metadata_snapshot, context));
+        }
+    }
+    else*/
+    {
+        /// For queries like `SELECT count(*) FROM tumble(table, now(), 5s) GROUP BY window_end` don't have required column from table.
+        /// We will need add one
+        Block header;
+        if (!column_names.empty())
+            header = storage_snapshot->getSampleBlockForColumns(column_names);
+        else
+            header = storage_snapshot->getSampleBlockForColumns({ProtonConsts::RESERVED_EVENT_TIME});
+
+        auto offsets = getOffsets(query_info.seek_to_info, shards_to_query);
+        assert(offsets.size() == shards_to_query.size());
+
+        for (auto [shard, offset] : std::ranges::views::zip(shards_to_query, offsets))
+            pipes.emplace_back(
+                std::make_shared<KafkaSource>(
+                    *this,
+                    header,
+                    storage_snapshot,
+                    consumer,
+                    topic_ptr,
+                    shard,
+                    offset,
+                    max_block_size,
+                    external_stream_counter,
+                    context));
+    }
+
+    LOG_INFO(
+        logger,
+        "Starting reading {} streams by seeking to {} in dedicated resource group",
+        pipes.size(),
+        query_info.seek_to_info->getSeekTo());
+
+    auto pipe = Pipe::unitePipes(std::move(pipes));
+    auto min_threads = context->getSettingsRef().min_threads.value;
+    if (min_threads > shards_to_query.size())
+        pipe.resize(min_threads);
+
+    return pipe;
+}
+
+RdKafka::Producer & Kafka::getProducer()
+{
+    if (producer)
+        return *producer;
+
+    std::scoped_lock lock(producer_mutex);
+    /// Check again in case of losing the race
+    if (producer)
+        return *producer;
+
+    auto producer_ptr = std::make_unique<RdKafka::Producer>(rd_kafka_conf_dup(conf.get()), settings->poll_waittime_ms.value, logger);
+    producer.swap(producer_ptr);
+
+    return *producer;
+}
+
+RdKafka::Topic & Kafka::getProducerTopic()
+{
+    if (producer_topic)
+        return *producer_topic;
+
+    std::scoped_lock lock(producer_mutex);
+    /// Check again in case of losing the race
+    if (producer_topic)
+        return *producer_topic;
+
+    auto topic_ptr = std::make_unique<RdKafka::Topic>(*getProducer().getHandle(), topic());
+    producer_topic.swap(topic_ptr);
+
+    return *producer_topic;
+}
```

Comment:
it seems the topic will be created at the first time it's used.
we should create it during initialize the external stream  - Kafka

Reply:
For producer or consumer, we can initialize it lazily

Reply:
If we only create an external stream, but don't read or write it.  the topic should be created, and there is not need consumer and producer

Reply:
yes, the topic just is a handler with reference count.